Assets/Scripts/Inventory/Inventory.cs

```diff
@@ -2,65 +2,148 @@
+            }
+        }
 
-        if(slot != null) // Item already exists
+        [SerializeField]
+        private List<ItemSlot> Items ; // Holds the player's collected items
+        public List<ItemSlot> GetInventoryItemSlots()
+        {
+            return Items;
+        }
+
+        public void AddNewItemToSlot(ItemSlot slot, Item item, int count)
+        {
+            // If the item slot does not contain an item yet
+            if (slot.item == null)
+            {
+                // ASSMPUTION IS THAT WE ARE ALWAYS ADDING COUNT HERE
+                slot.item = item;
+                slot.ItemCount = count;
+                return;
+            }
+            Debug.LogError("Attempting to add item to a non null item slot");
+        }
+
+        // Given an item slot, change its count
+        public void ModifyItemSlot(ItemSlot slot, int count)
         {
             // Ensure that infinite is not on when we want to modify item count
             if (!this.InfiniteItem)
             {
                 // Ensure we don't go over max
-                if(MaxNumberOfItem < (slot.ItemCount + count))
+                if (MaxNumberOfItem > (slot.ItemCount + count))
                 {
                     // Add count
                     slot.ItemCount += count;
                 }
+                // Ensure we don't go below 0
+                else if (0 >= (slot.ItemCount + count))
+                {
+                    // Else we just go to 0
+                    slot.ItemCount = 0;
+                }
             }
         }
-        else // Item does not exist
+
+        public void ModifyItemSlot(Item item, int count)
         {
-            // Add to list
-            Items.Add(new ItemSlot(item, 2));
+            ItemSlot slot = FindItem(item.thisItemType);
+
+            if (slot != null)
+                ModifyItemSlot(slot, count);
+            else
+            {
+                // A little inefficient
+                if(FindEmptySlot(out slot))
+                {
+                    AddNewItemToSlot(slot, item, count);
+                }
+                else
+                {
+                    // No empty slots found
+                    Debug.LogWarning("No empty slots found to add item");
+                }
+            }
         }
-    }
 
-    public ItemSlot FindItem(Item.ItemType itemType)
-    {
         // Go through the inventory list to find if we already have the same type
-        foreach (ItemSlot thisSlot in Items)
+        public ItemSlot FindItem(Item.ItemType itemType)
+        {
+            foreach (ItemSlot thisSlot in Items)
+            {
+                if (thisSlot.item?.thisItemType == itemType)
+                {
+                    return thisSlot;
+                }
+            }
+            return null;
+        }
+
+        // Returns null when no slots are available
+        public ItemSlot FindItemOrEmptySlot(Item.ItemType itemType)

```

Comment:
This confuse me a little bit. We have a method `FindItem` and we have a method `FindEmptySlot`. Why do we need another method `FindItemOrEmptySlot`?

Reply:
Mostly experimental to see how the code can change. If its a big issue I can remove it.